bob-apps/bin/bobc.rs

```diff
@@ -1,71 +1,365 @@
+        }
+        match keys {
+            KeyPattern::Single(val) => {
+                vec![(*val, RE.replace(self.inner, val.to_string()).into_owned())]
+            }
+            KeyPattern::Range(r) => {
+                r.clone() // idk how to use map Range without moving out a value
+                    .map(|n| (n, RE.replace(self.inner, n.to_string()).into_owned()))
+                    .collect()
+            }
+            KeyPattern::Multiple(v) => v
+                .into_iter()
+                .map(|n| (*n, RE.replace(self.inner, n.to_string()).into_owned()))
+                .collect(),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum KeyPattern {
+    Single(u64),
+    Range(RangeInclusive<u64>),
+    Multiple(Vec<u64>),
+}
+
+#[derive(Debug)]
+struct KeyName {
+    key: Vec<u8>,
+    name: String,
+}
+
 #[tokio::main]
 async fn main() {
     env_logger::builder().filter_level(LevelFilter::Info).init();
     let matches = get_matches();
-    if let (sc, Some(sub_mathes)) = matches.subcommand() {
-        let key = get_key_value(sub_mathes);
-        match sc {
-            "put" => {
-                let value = sub_mathes.value_of("size").unwrap();
-                let size = value.parse().expect("size must be usize");
-                let addr = sub_mathes
-                    .value_of("uri")
-                    .expect("has default value")
-                    .parse()
-                    .expect("wrong format of url");
-                info!("PUT key: \"{:?}\" size: \"{}\" to \"{}\"", key, size, addr);
-                put(key, size, addr).await;
+    let app_args = AppArgs::from_matches(&matches);
+
+    let addr = NetConfig::from_args(&app_args).get_uri();
+    match app_args.subcommand {
+        PUT_SC => {
+            let keys_names = match app_args.file_pattern {
+                Some(FilePattern::WithRE(re, dir)) => {
+                    // maybe create filenames from pattern AND keys and try to send them?
+                    if app_args.key_pattern != None {
+                        return error!("Either file pattern or key argument must be used");
+                    }
+                    prepare_put_from_pattern(&re.get_regex(), &dir, app_args.keysize).await
+                }
+                Some(FilePattern::WithoutRE(path)) => match app_args.key_pattern {
+                    None => {
+                        return error!("Key arg is required if not using file pattern");
+                    }
+                    Some(p) => match p {
+                        KeyPattern::Single(key) => {
+                            let key = get_key_value(key, app_args.keysize);
+                            vec![KeyName {
+                                key,
+                                name: path.to_string(),
+                            }]
+                        }
+                        _ => return error!("Multiple keys are not allowed without pattern"),
```

Comment:
Actually, I don't see a problem in allowing multiple keys (single file just will be written for every specified key)

Reply:
But how do we determine file names?

Reply:
For `PUT_SC` it is possible but not for `GET_SC`.

Reply:
Done: [8f88c63](https://github.com/qoollo/bob/pull/562/commits/8f88c638c84a18e4e516452f3467f3c58edc4f56)